src/type/dispatcher.js

```diff
@@ -0,0 +1,102 @@
+// @flow strict
+
+import { $$asyncIterator } from 'iterall';
+
+import { type Path, pathToArray } from '../jsutils/Path';
+import { type PromiseOrValue } from '../jsutils/PromiseOrValue';
+import isPromise from '../jsutils/isPromise';
+
+import { GraphQLError } from '../error/GraphQLError';
+
+export class Dispatcher {
+  _patches: Array<Promise<{| value: ExecutionPatchResult, done: boolean |}>>;
+
+  constructor() {
+    this._patches = [];
+  }
+
+  execute(fn: () => PromiseOrValue<mixed>, errors: Array<GraphQLError>) {
+    try {
+      return fn();
+    } catch (error) {
+      errors.push(error);
+      return null;
+    }
+  }
+
+  add(
+    label: string,
+    path: Path | void,
+    fn: () => PromiseOrValue<mixed>,
+    errors: Array<GraphQLError>,
+  ) {
+    this._patches.push(
+      Promise.resolve(this.execute(fn, errors)).then(data => {
+        if (isPromise(data)) {
+          return data.then(undefined, error => ({
+            value: {
+              data: null,
+              path: pathToArray(path),
+              label,
+              errors: [error],
+            },
+            done: false,
+          }));
+        }
+        return {
+          value: {
+            data,
+            path: pathToArray(path),
+            label,
+            ...(errors && errors.length > 0 ? { errors } : {}),
+          },
+          done: false,
+        };
+      }),
+    );
+  }
+
+  get(): AsyncIterable<ExecutionPatchResult> | null {
+    if (this._patches.length === 0) {
+      return null;
+    }
+    const results = this._patches;
+
+    function race(promises) {
+      return new Promise(resolve => {
+        promises.forEach((promise, index) => {
+          promise.then(result => {
+            resolve({ result, index });
+          });
+        });
+      });
+    }
+
+    const getNext = promises => {
+      if (promises.length === 0) {
+        return Promise.resolve({ value: undefined, done: true });
+      }
+      return race(promises).then(({ result, index }) => {
+        promises.splice(index, 1);
+        return result;
+      });
+    };
+
+    return ({
+      next() {
+        return getNext(results);
+      },
+      [$$asyncIterator]() {
+        return this;
+      },
+    }: any);
+  }
+}
+
+export type ExecutionPatchResult = {
+  errors?: $ReadOnlyArray<GraphQLError>,
+  data?: mixed | null,
+  path: $ReadOnlyArray<string | number>,
+  label: string,
+  ...
```

Comment:
@Keweiqu thanks for the feedback! This is very helpful information. Do you think `is_final` in the extensions field is something that needs to be part of the spec for this feature, or is that more of a network implementation detail? In our testing, the patches were sent over a streaming HTTP connection. We were able to signal to Relay that there are no more patches by detecting the HTTP connection has been closed.

I will add some more tests to make sure the `AsyncIterator` is closed correctly for queries that contain multiple `@stream`s and `@defer`s.

Reply:
@robrichard good question. For the purpose of this PR, my inclination is to aim for only the absolutely essential features, because it's easier to add than to remove. 

If we can signal to Relay through detecting a closed HTTP connection without relying on the "extension" field, and @stream/@defer proves to work fine. Let's start with it as the initial step.

We plan to initiate the discussion of evaluating whether to formalize @defer and @stream as a spec and how to best go about it in the next graphql working group meeting. From there on, we will have more information on whether `is_final` will be beneficial to the open source community at large. It will also help inform whether it's a good idea to include it in the reference implementation of graphql-js.